sw/host/opentitanlib/src/io/jtag.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use std::rc::Rc;
+use structopt::StructOpt;
+use thiserror::Error;
+
+use crate::app::TransportWrapper;
+use crate::dif::lc_ctrl::{DifLcCtrlState, LcCtrlReg};
+use crate::impl_serializable_error;
+
+#[derive(Debug, StructOpt)]
+pub struct JtagParams {
+    #[structopt(long, help = "OpenOCD binary path.", default_value = "openocd")]
+    pub openocd: String,
+
+    #[structopt(long, help = "Path to OpenOCD JTAG adapter config file.")]
+    openocd_adapter_config: String,
+}
+
+impl JtagParams {
+    pub fn create(&self, transport: &TransportWrapper) -> Result<Rc<dyn Jtag>> {
+        let jtag = transport.jtag(&self.openocd, &self.openocd_adapter_config)?;
+        Ok(jtag)
+    }
+}
+
+/// Errors related to the JTAG interface.
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum JtagError {
+    #[error("JTAG timeout")]
+    Timeout,
+    #[error("JTAG busy")]
+    Busy,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(JtagError);
+
+/// Represents an OpenOCD command sent over JTAG.
+pub enum OpenOcdCmd {
+    ReadLcCtrlReg(LcCtrlReg),
+    WriteLcCtrlReg(LcCtrlReg, u32),
+    // ReadCpuReg(IbexReg),
+    // WriteCpuReg(IbexReg, u32),
+    ReadMemory { addr: u32 },
+    WriteMemory { addr: u32, value: u32 },
+    Halt,
+    Resume,
+    Shutdown,
+    Version,
+}
+
+/// Represents a (higher complexity) OpenOCD command sequence.
+pub enum OpenOcdCmdSeq {
+    /// Exits with a non-zero status when the register does not contain the
+    /// given value.
+    AssertLcCtrlRegEq(LcCtrlReg, u32),
+
+    /// Polls the lc_ctrl register a finite number of times until its value
+    /// matches the expectation.
+    PollUntilLcCtrlRegEq(LcCtrlReg, u32),
+
+    /// Performs a lifecycle state transition to the target lifecycle state.
+    PerformLcTransition(DifLcCtrlState),
+}
+
+/// A trait which represents a JTAG interface.
+pub trait Jtag {
```

Comment:
This interface seems to assume that you can run interactive commands with OpenOCD which you can do with telnet. In this case, isn't it possible/better to also read the result of the command? For example to read a register, just execute "riscv dmi_read 13" and read the result back. This would allow the Jtag to be less about OpenOCD and more about the operations, and to be more state-less. Otherwise you'll end up executing things like "set my_reg [riscv dmi_read 13]" and it all becomes super specific to OpenOCD and also very stateful.
What do you think?

Reply:
Yes, agreed, I think it makes more sense to update this API to be more action centric.

The statelessness will depend on if the OpenOcdServer struct you create, that I assume you will instantiate somewhere within the a `CW310Jtag` struct you will create (similar to the `CW310Spi` [struct](https://cs.opensource.google/opentitan/opentitan/+/master:sw/host/opentitanlib/src/transport/cw310/spi.rs;drc=02b2c0a6067cc654457be84ec5faa6ea8e8bdcbc;l=14)), will automatically spin-up the openocd server process on construction and tear it down on destruction? or it will wait to be instructed to be spun up / down by API calls.

The reason I was leaning toward the latter, is that not everyone has a JTAG adapter yet that has a CW310 board, so I thought it might be nice to leave it up to the test to manage the openocd server state through invoking API calls (then we can essentially we can continue to tag tests with the Bazel "jtag" tag to indicate tests that require this extra HW resource). Otherwise everytime a CW310 transport is instantiated with opentitanlib, we need to ensure there is something for it to connect to, or be just not care if it connects to anything.

Any thoughts on this?

Reply:
I was thinking that either we start the server on the first API call, or we have some explicit connect()/disconnect() methods in Jtag that abstract the notion of connecting to the Jtag chain. This way jtag/OpenOCD is not required for the non-jtag operations.
There is also the issue of how to handle the tap straps: if we are going to support dynamic changes of the straps, then the openocd server might need to be restarted (I am not sure if want to support dynamically changing the taps and configuration?) so it makes sense to explicitely manage the jtag connection state.